cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go

```diff
@@ -90,49 +91,61 @@ func UploadConfiguration(cfg *kubeadmapi.InitConfiguration, client clientset.Int
 		return err
 	}
 
-	err = apiclient.CreateOrMutateConfigMap(client, &v1.ConfigMap{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      kubeadmconstants.KubeadmConfigConfigMap,
-			Namespace: metav1.NamespaceSystem,
-		},
-		Data: map[string]string{
-			kubeadmconstants.ClusterConfigurationConfigMapKey: string(clusterConfigurationYaml),
-			kubeadmconstants.ClusterStatusConfigMapKey:        string(clusterStatusYaml),
-		},
-	}, func(cm *v1.ConfigMap) error {
-		// Upgrade will call to UploadConfiguration with a modified KubernetesVersion reflecting the new
-		// Kubernetes version. In that case, the mutation path will take place.
-		cm.Data[kubeadmconstants.ClusterConfigurationConfigMapKey] = string(clusterConfigurationYaml)
-		// Mutate the ClusterStatus now
-		return mutateClusterStatus(cm, func(cs *kubeadmapi.ClusterStatus) error {
-			// Handle a nil APIEndpoints map. Should only happen if someone manually
-			// interacted with the ConfigMap.
-			if cs.APIEndpoints == nil {
-				return errors.Errorf("APIEndpoints from ConfigMap %q in the %q Namespace is nil",
-					kubeadmconstants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
-			}
-			cs.APIEndpoints[cfg.NodeRegistration.Name] = cfg.LocalAPIEndpoint
-			return nil
+	err = wait.PollImmediate(kubeadmconstants.APICallRetryInterval, kubeadmconstants.APICallWithWriteTimeout, func() (bool, error) {
+		err = apiclient.CreateOrMutateConfigMap(client, &v1.ConfigMap{
```

Comment:
so this function already has retries:
https://github.com/kubernetes/kubernetes/blob/5708511499fe500ae3b4bbd40204cef382f652e8/cmd/kubeadm/app/util/apiclient/idempotency.go#L63-L85

but my search tells me that it's only used inside this file.

so i think we should remove the RetryOnConflict and backoff from CreateOrMutateConfigMap and make it one shot like the rest of the functions in idempotency.go

then move the retries on the side of the caller, like it's done in your current diff.

i don't think we should use the same semantics as retryOnConflict and just retry on all types of errors...:
https://github.com/kubernetes/kubernetes/blob/8d247629c38228ac18513612024c527f58ed9c57/staging/src/k8s.io/client-go/util/retry/util.go#L104

Reply:
@neolit123 should we also the retry of `MutateConfigMap()`, and add the retry in the caller:

https://github.com/kubernetes/kubernetes/blob/f1a86101176a24e8097a28cdd9a664e1cde625f2/cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go#L48

Reply:
CreateOrMutateConfigMap has retries which then calls MutateConfigMap which also has retries. this is quite messy.

we have to backport this PR to 1.17 and 1.18 so we have to keep the change minimal, but looks like the existing code is already quite messy, so my vote is yes - let's change MutateConfigMap too.